content/blogs/gui-setup-part3-first-testing-problems.md

```diff
@@ -0,0 +1,267 @@
+The exception could be, that if the next file is not in the order, one restarts finding from the beginning of the CAB archive.
+This would be the slow code path, but not actually taken in real life due to the reasons from above.
+
+This is exactly what the text-mode USETUP installer does, and why ReactOS installation time is there blazing fast in comparison.
+
+Additionally, it appears, after some further investigation (suggested by Mark Jansen), that each read operation from the CAB file, implemented with a `fdi->read()` callback (registered when initializing the fdi context with the `FDICreate()` function, done by setupapi), that does a `ReadFile()` call.
+Meaning, basically, that each data read from the CAB triggers a file I/O operation to disk, making the whole operation slower than it could be.
+This could be alleviated by adding ReactOS-specific code in setupapi, to memory-map the CAB archive instead so that reading from it can be done directly via memory reads.
+Such a thing is also done by the text-mode USETUP.
+
+#### The current workaround
+
+To alleviate the problems described above, the simplest workaround is to just use USETUP's file-copy and extraction functions. With minimal adaptations the existing code can be easily adapted for usage in the GUI installer. File copy works and finishes speedily.
+
+{{< gallery >}}
+{{< figure link="/img/blogs/hbelusca_gui-setup/guisetup_filecopy_finish.png" src="/img/blogs/hbelusca_gui-setup/guisetup_filecopy_finish.png" caption="Last file being copied. The directory listing shows that the files are there.">}}
+{{< /gallery >}}
+
+
+### 3rd bug
+
+At the end of the installation, the ReactOS GUI installer creates the necessary registry hives (SYSTEM, SOFTWARE, SAM and SECURITY) in their correct place in the installation directory (`%SystemRoot%\system32\config`).
+Using NT APIs, they are created, filled for one part with default settings and for another part with installation-specific settings, then flushed back to disk.
+
+Notice that these created registry hives are separate from the system root ones being used by the LiveCD.
+In the LiveCD (running in the so-called "WinPE" or "MiniNT" mode), the system root hives are mounted with shared (read-only) access, because they reside on a read-only medium usually, or in network boot they are on an (external) shared network.
+Thus for these LiveCD hives, any writes made to them (and any flush to disk) is a no-operation.
+
+{{< gallery >}}
+{{< figure link="/img/blogs/hbelusca_gui-setup/guisetup_create_registry.png" src="/img/blogs/hbelusca_gui-setup/guisetup_create_registry.png" caption="Creating the registry hives">}}
+{{< /gallery >}}
+
+Other hives (those that can be loaded manually) can be loaded with full access because they can reside elsewhere than on a read-only medium (and can be on the local computer).
```

Comment:
```suggestion
Other hives (those that could be loaded manually) can be loaded with full access because they may reside elsewhere than on a read-only medium (and can be on the local computer).
```